### PR TITLE
[ci] resolve shellcheck errors in .ci/test_r_packages.sh

### PR DESCRIPTION
**Pull request by Kunal-Singh-Dadhwal**
Contributes to #6498

So here i rectified the errors in test_r_packages.sh and also checked it with shellcheck

### CODE REVIEW

**Comment by Kunal-Singh-Dadhwal** [comment]
@microsoft-github-policy-service agree

**Comment by Kunal-Singh-Dadhwal** [comment]
@microsoft-github-policy-service agree

File: .ci/test_r_package.sh [hunk 1]
```diff
@@ -70,17 +70,17 @@ if [[ $OS_NAME == "linux" ]]; then
...
     fi
     if [[ $INSTALL_CMAKE_FROM_RELEASES == "true" ]]; then
         curl -O -L \
-            https://github.com/Kitware/CMake/releases/download/v3.25.1/cmake-3.25.1-linux-${ARCH}.sh \
+            https://github.com/Kitware/CMake/releases/download/v3.25.1/cmake-3.25.1-linux-"${ARCH}".sh \
         || exit 1
 
         sudo mkdir /opt/cmake || exit 1
-        sudo sh cmake-3.25.1-linux-${ARCH}.sh --skip-license --prefix=/opt/cmake || exit 1
+        sudo sh cmake-3.25.1-linux-"${ARCH}".sh --skip-license --prefix=/opt/cmake || exit 1
```
Comment (jameslamb) [inline .ci/test_r_package.sh:83]:
```suggestion
        sudo sh "cmake-3.25.1-linux-${ARCH}.sh" --skip-license --prefix=/opt/cmake || exit 1
```

Please don't put double quotes inside of paths like this... let's quote the entire path.

File: .ci/test_r_package.sh [hunk 2]
```diff
@@ -70,17 +70,17 @@ if [[ $OS_NAME == "linux" ]]; then
...
             -y \
-                autoconf=$(cat R-package/AUTOCONF_UBUNTU_VERSION) \
+                autoconf="$(cat R-package/AUTOCONF_UBUNTU_VERSION)" \
                 automake \
                 || exit 1
     fi
     if [[ $INSTALL_CMAKE_FROM_RELEASES == "true" ]]; then
         curl -O -L \
-            https://github.com/Kitware/CMake/releases/download/v3.25.1/cmake-3.25.1-linux-${ARCH}.sh \
+            https://github.com/Kitware/CMake/releases/download/v3.25.1/cmake-3.25.1-linux-"${ARCH}".sh \
```
Comment (jameslamb) [inline .ci/test_r_package.sh:79]:
```suggestion
            "https://github.com/Kitware/CMake/releases/download/v3.25.1/cmake-3.25.1-linux-${ARCH}.sh" \
```

File: .ci/test_r_package.sh [hunk 3]
```diff
@@ -100,15 +100,15 @@ if [[ $OS_NAME == "macos" ]]; then
     sudo tlmgr --verify-repo=none update --self || exit 1
     sudo tlmgr --verify-repo=none install inconsolata helvetic rsfs || exit 1
 
-    curl -sL ${R_MAC_PKG_URL} -o R.pkg || exit 1
+    curl -sL "${R_MAC_PKG_URL}" -o R.pkg || exit 1
     sudo installer \
-        -pkg $(pwd)/R.pkg \
+        -pkg "$(pwd)"/R.pkg \
```
Comment (jameslamb) [inline .ci/test_r_package.sh:105]:
```suggestion
        -pkg "$(pwd)/R.pkg" \
```

File: .ci/test_r_package.sh [hunk 4]
```diff
@@ -236,12 +236,10 @@ if [[ $used_correct_r_version -ne 1 ]]; then
     echo "Unexpected R version was used. Expected '${R_VERSION}'."
     exit 1
 fi
-
 if [[ $R_BUILD_TYPE == "cmake" ]]; then
-    passed_correct_r_version_to_cmake=$(
```
Comment (jameslamb) [inline .ci/test_r_package.sh:241]: what is the purpose for these changes? Can you please explain them?

Providing the text of the exact error(s) from `shellcheck` that you're attempting to address would be helpful.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for your interest in LightGBM!

I've added a link to the issue I suspect you found this work from (#6498). Please do that in the future, it helps everyone involved to understand why you're proposing these changes.

Please click through the links to CI jobs and fix the failures reported there. You can `@` me help if you need it.

Also... in the future, when you contribute to open source projects on GitHub, don't use the `master` / `main` branch of your fork. Create a new branch with your changes. Now that there have been some changes merged to `master` here, when you merge those into your fork's `master` branch, it'll have a different history than this repo's `master` branch, which will cause problems if you create another pull request in the future.

**Comment by jameslamb** [comment]
Please also merge the latest state of `master` into this branch.

**Comment by Kunal-Singh-Dadhwal** [comment]
Actually this was my first open source contribution. Thank you for your feedback and guidance! I appreciate your insights and will make sure to follow your suggestions for future contributions.

File: .ci/test_r_package.sh (hunk 4 above)
Comment (Kunal-Singh-Dadhwal) [reply .ci/test_r_package.sh:241]: In testing.sh line 241:
    passed_correct_r_version_to_cmake=$(
    ^-- SC2034 (warning): passed_correct_r_version_to_cmake appears unused. Verify use (or export if used externally).

And after declaring the variable as export this was the error
In testing.sh line 241:
    export passed_correct_r_version_to_cmake=$(
           ^-- SC2155 (warning): Declare and assign separately to avoid masking return values.

So then i declared passed_correct_r_version_to_cmake seperately and then made the count variable and initialized passed_correct_r_version_to_cmake to count and check for the version

Comment (jameslamb) [reply .ci/test_r_package.sh:241]: Thanks. Note that I've edited your post, to make the difference between code, output from code, and your own words clearer. You may want to review https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax.

---

This fix does not look correct to me. It looks like the actual issue is that this check is storing a value in a variable called `passed_correct_r_version_to_cmake`, but then the `if` statement below is referencing a variable called `used_correct_r_verrsion`.

Please revert your changes in this part of the code, and instead change this line:

https://github.com/microsoft/LightGBM/blob/fd8f79944955cf83fcefb9ed598b9e0bcd07ad09/.ci/test_r_package.sh#L245

to use `passed_correct_r_version_to_cmake`

**Comment by Kunal-Singh-Dadhwal** [comment]
> Please also merge the latest state of `master` into this branch
Should i make my own branch and push these changes

**Comment by jameslamb** [comment]
> this was my first open source contribution. 

Oh great, thanks for choosing LightGBM for that! We're happy to have you.

In the future, when you make a contribution (especially to a project you've never worked on before), please provide more information in the pull request title / description than "resolve errors". For example, can you please update the description here with the full text of all the errors you're trying to address?

That'd help reviewers to understand specifically what you are attempting to fix, and would help answer the important question "why should we merge this? how does this improve the project?".

**Comment by jameslamb** [comment]
> Should i make my own branch and push these changes

Not for this PR. **After this PR is merged**, you should do the following if you intend to contribute here again:

* delete your LigthGBM fork here on GitHub
* recreate it

And then in the future if you contribute here, you should work from a new branch on your fork for every pull request, like this:

```shell
git checkout -b ci/shellcheck
```

File: .ci/test_r_package.sh (hunk 4 above)
Comment (Kunal-Singh-Dadhwal) [reply .ci/test_r_package.sh:241]: Changed the variable name. Thank you for the link to the github docs

Comment (jameslamb) [reply .ci/test_r_package.sh:241]: I think you've misunderstood. Here in this part of the code, the variable name should be changed **and nothing else should**.

All of the other changes you've made (like marking `passed_correct_r_version_to_cmake` for export), should be reverted. Please revert them.

File: .ci/test_r_package.sh [hunk 5]
```diff
@@ -167,45 +167,43 @@ elif [[ $R_BUILD_TYPE == "cran" ]]; then
...
-    R CMD check ${PKG_TARBALL} \
-        --as-cran \
-        --run-donttest \
-    || check_succeeded="no"
-) &
+if R CMD check "${PKG_TARBALL}" --as-cran --run-donttest; then
+    export check_r_package="yes"
+else
+    export check_r_package="no"
+fi
```
Comment (jameslamb) [inline .ci/test_r_package.sh:206]: I don't understand this change (and then later putting `check_succeeded` below). What is the purpose of this? Can you please provide the exact text of the error(s) this is attempting to fix?

File: .ci/test_r_package.sh [hunk 6]
```diff
@@ -167,45 +167,43 @@ elif [[ $R_BUILD_TYPE == "cran" ]]; then
...
             "/rchk/packages/${PKG_TARBALL}" \
-        2>&1 > ${RCHK_LOG_FILE} \
+         > ${RCHK_LOG_FILE} 2>&1 \
         || (cat ${RCHK_LOG_FILE} && exit 1)
         cat ${RCHK_LOG_FILE}
 
         # the exceptions below are from R itself and not LightGBM:
         # https://github.com/kalibera/rchk/issues/22#issuecomment-656036156
-        exit $(
+        exit "$(
```
Comment (jameslamb) [inline .ci/test_r_package.sh:182]: What error did `shellcheck` raise here? Did it perhaps actually want you to put quotes around `RCHK_LOG_FILE` on the line below, instead?

File: .ci/test_r_package.sh [hunk 7]
```diff
@@ -167,45 +167,43 @@ elif [[ $R_BUILD_TYPE == "cran" ]]; then
     if [[ "${TASK}" == "r-rchk" ]]; then
         echo "Checking R package with rchk"
         mkdir -p packages
-        cp ${PKG_TARBALL} packages
+        cp "${PKG_TARBALL}" packages
         RCHK_LOG_FILE="rchk-logs.txt"
         docker run \
-            -v $(pwd)/packages:/rchk/packages \
+            -v "$(pwd)"/packages:/rchk/packages \
```
Comment (jameslamb) [inline .ci/test_r_package.sh:173]:
```suggestion
            -v "$(pwd)/packages":/rchk/packages \
```

The full path, please.

File: .ci/test_r_package.sh [hunk 8]
```diff
@@ -147,7 +147,7 @@ elif [[ $R_BUILD_TYPE == "cran" ]]; then
     # on Linux, we recreate configure in CI to test if
     # a change in a PR has changed configure.ac
     if [[ $OS_NAME == "linux" ]]; then
-        ${BUILD_DIRECTORY}/R-package/recreate-configure.sh
+        "${BUILD_DIRECTORY}"/R-package/recreate-configure.sh
```
Comment (jameslamb) [inline .ci/test_r_package.sh:150]:
```suggestion
        "${BUILD_DIRECTORY}/R-package/recreate-configure.sh"
```

Quote the full path on every one of these, please.

**Review by jameslamb — CHANGES_REQUESTED** [review]

File: .ci/test_r_package.sh (hunk 6 above)
Comment (Kunal-Singh-Dadhwal) [reply .ci/test_r_package.sh:182]: In testing.sh line 182:
        exit $(
             ^-- SC2046 (warning): Quote this to prevent word splitting.

File: .ci/test_r_package.sh (hunk 5 above)
Comment (Kunal-Singh-Dadhwal) [reply .ci/test_r_package.sh:206]: In testing.sh line 207:
    || check_succeeded="no"
       ^-------------^ SC2030 (info): Modification of check_succeeded is local (to subshell caused by (..) group).

Comment (jameslamb) [reply .ci/test_r_package.sh:206]: This change is just not correct. We might be losing the assignment of `check_succeeded` (that looks like a genuine bug), but it's intentional that this is run in a background process (using `&`), for the reasons described in the comment directly below these lines.

It's possible that we could remove this pattern of running `R CMD check` in a background task and refactor this a bit, but if you are not familiar with R or that command then it's going to be difficult for you to do that.

For the purpose of this PR, please just revert these changes related to this `R CMD check` run. I'll try resolving that in a separate PR.

File: .ci/test_r_package.sh (hunk 4 above)
Comment (Kunal-Singh-Dadhwal) [reply .ci/test_r_package.sh:241]: Did the change. Sorry for misunderstanding

File: .ci/test_r_package.sh [hunk 9]
```diff
@@ -148,7 +148,7 @@ elif [[ $R_BUILD_TYPE == "cran" ]]; then
     # a change in a PR has changed configure.ac
     if [[ $OS_NAME == "linux" ]]; then
         ./R-package/recreate-configure.sh
-
+
```
Comment (jameslamb) [inline .ci/test_r_package.sh:151]:
```suggestion

```

Please revert this additional whitespace added.

File: .ci/test_r_package.sh [hunk 10]
```diff
@@ -236,13 +236,12 @@ if [[ $used_correct_r_version -ne 1 ]]; then
     echo "Unexpected R version was used. Expected '${R_VERSION}'."
     exit 1
 fi
-
 if [[ $R_BUILD_TYPE == "cmake" ]]; then
-    passed_correct_r_version_to_cmake=$(
+     passed_correct_r_version_to_cmake=$(
```
Comment (jameslamb) [inline .ci/test_r_package.sh:240]:
```suggestion
    passed_correct_r_version_to_cmake=$(
```

**Review by jameslamb — CHANGES_REQUESTED** [review]

File: .ci/test_r_package.sh (hunk 6 above)
Comment (jameslamb) [reply .ci/test_r_package.sh:181]: hmmmm ok. I think I disagree with `shellcheck` on this one, and it's a bit strange to see a `"` at the beginning of such a multiline statement.

Can you please revert the quotes for this statement, and instead add a comment directly above (indented over to match the `exit`) like this?

```text
# shellcheck disable=SC2046
```

File: .ci/test_r_package.sh (hunk 5 above)
Comment (Kunal-Singh-Dadhwal) [reply .ci/test_r_package.sh:205]: check_r_package="yes" && { R CMD check "${PKG_TARBALL}" --as-cran --run-donttest || check_r_package="no"; } &
 can we do this ?

File: .ci/test_r_package.sh [hunk 11]
```diff
@@ -148,7 +148,6 @@ elif [[ $R_BUILD_TYPE == "cran" ]]; then
     # a change in a PR has changed configure.ac
     if [[ $OS_NAME == "linux" ]]; then
         ./R-package/recreate-configure.sh
-
```
Comment (jameslamb) [inline .ci/test_r_package.sh:151]: Please add this empty line back, unless it's something `shellcheck` was complaining about.

File: .ci/test_r_package.sh (hunk 5 above)
Comment (jameslamb) [reply .ci/test_r_package.sh:205]: No, it's more complex than that and involves changing more than just this line. Please, for the purpose of this PR, just do not touch this part of this script.

File: .ci/test_r_package.sh [hunk 12]
```diff
@@ -219,7 +220,9 @@ done
 
 echo "R CMD check build logs:"
 BUILD_LOG_FILE=lightgbm.Rcheck/00install.out
-cat ${BUILD_LOG_FILE}
+cat "${BUILD_LOG_FILE}"
+
+check_succeeded=$check_r_package
```
Comment (jameslamb) [inline .ci/test_r_package.sh:225]:
```suggestion
```

Please remove these (they'd been added as part of the diff I asked you to revert in https://github.com/microsoft/LightGBM/pull/6501#discussion_r1649266385

**Review by jameslamb — CHANGES_REQUESTED** [review]
This is getting closer!

Please look through the diff at https://github.com/microsoft/LightGBM/pull/6501/files to identify whitespace-only changes, and please revert those.

I see two. This one on line 151 where you've added back these spaces I asked you to remove:

image

And this one, where you've removed an entire line:

image

If you're unsure how to find those extra space characters, consider changing the settings in your editor to display whitespace characters.

**Comment by Kunal-Singh-Dadhwal** [comment]
@jameslamb removed whitespaces

**Comment by jameslamb** [comment]
Please put this line back:

image

**Comment by jameslamb** [comment]
Thanks, but it seems that the most recent commit you pushed re-introduced those extra space characters. I think maybe you  aren't aware of how to identify space-only characters when adding text files.

No problem, in some editors they're hidden! To get this PR merged, I just pushed a commit removing those. You can view it here: https://github.com/microsoft/LightGBM/pull/6501/commits/e1c71cc2257a9292e8b2e2a4c18020f8a7d6e752

You may want to turn on display of whitespace characters in your editor. Here's what this file looks like to me in VS Code (note that dots for spaces): 

image

**Comment by Kunal-Singh-Dadhwal** [comment]
Will check thoroughly the next time I make any contribution. Thanks for your patience and guidance

**Review by jameslamb — APPROVED** [review]
Thanks very much for this!

I've pushed a few commits here to finish this.

We'd love your help with more `shellcheck` things to make progress on #6498 if it interests you. If you do submit future pull requests ("PR"), please consider the suggestions I've given in this review:

* do not use the `master` branch... create a new branch on your fork, from `master`
* link to #6498 in the description of the PR
* add the exact text of the errors that your PR seeks to fix, in the description of the PR, formatted as a code block following https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

**Comment by jameslamb** [comment]
@Kunal-Singh-Dadhwal it looks like you deleted your fork before this was merged, so now your work won't actually be added to LightGBM. I strongly suspect that was a mistake.

If it was and you are still interested in contributing these changes, please open a new pull request.

**Comment by Kunal-Singh-Dadhwal** [comment]
@jameslamb Yes it was a mistake Sorry ,I would make another PR by today
